File: app-server/src/test/java/com/github/arhor/simple/expense/tracker/data/repository/NotificationRepositoryTest.java
package com.github.arhor.simple.expense.tracker.data.repository;

import lombok.val;

import java.time.LocalDateTime;
import java.time.temporal.ChronoUnit;
import java.util.function.Function;
import java.util.stream.IntStream;
import java.util.stream.Stream;

import org.junit.jupiter.api.Test;

import com.github.arhor.simple.expense.tracker.data.model.InternalUser;
import com.github.arhor.simple.expense.tracker.data.model.Notification;

import static java.util.stream.Collectors.toMap;
import static org.assertj.core.api.Assertions.assertThat;

class NotificationRepositoryTest extends RepositoryTestBase {

    @Test
    void should_return_all_notifications_where_user_id_is_one_of_the_passed_ids() {
        // given
        val usersIds = createUsersStream()
            .map(InternalUser::id)
            .toList();

        val notifications = usersIds.stream()
            .flatMap(this::createNotificationsStream)
            .map(notificationRepository::save)
            .collect(toMap(Notification::id, Function.identity()));

        // when
        val result = notificationRepository.findAllByTargetUserIdIn(usersIds);

        // then
        assertThat(result)
            .isNotEmpty()
            .allSatisfy(projection -> {
                val currentId = projection.id();
                val currentUserId = projection.targetUserId();

                assertThat(notifications.get(currentId))
                    .as(() -> "notification: " + currentId + ", user id: " + currentUserId)
                    .isNotNull()
                    .satisfies(
                        notification -> {
                            assertThat(projection.id())
                                .as("id")
                                .isEqualTo(notification.id());
                        },
                        notification -> {
                            assertThat(projection.targetUserId())
                                .as("targetUserId")
                                .isEqualTo(notification.targetUserId());
                        },
                        notification -> {
                            assertThat(projection.message())
                                .as("message")
                                .isEqualTo(notification.message());
                        },
                        notification -> {
                            assertThat(projection.severity())
                                .as("severity")
                                .isEqualTo(notification.severity());
                        }
                    );
            });
    }

    private Stream<InternalUser> createUsersStream() {
        return IntStream.range(0, 5).mapToObj(this::createPersistedTestUser);
    }

    private Stream<Notification> createNotificationsStream(final Long targetUserId) {
        return IntStream.range(0, 3).mapToObj(number ->
            Notification.builder()
                .targetUserId(targetUserId)
                .message("test-message-" + number)
                .severity("INFO")
                .timestamp(LocalDateTime.now().truncatedTo(ChronoUnit.MILLIS))
                .build()
        );
    }
}

File: app-server/src/test/kotlin/com/github/arhor/simple/expense/tracker/data/repository/NotificationRepositoryTest.kt
package com.github.arhor.simple.expense.tracker.data.repository

import com.github.arhor.simple.expense.tracker.data.model.InternalUser
import com.github.arhor.simple.expense.tracker.data.model.Notification
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import java.time.LocalDateTime
import java.time.temporal.ChronoUnit
import java.util.function.Consumer

internal class NotificationRepositoryTest : RepositoryTestBase() {

    @Test
    fun `should return all notifications where user id is one of the passed ids`() {
        // given
        val usersIds = createUsers().mapNotNull(InternalUser::id)

        val notifications = usersIds
            .flatMap { createNotifications(it) }
            .map(notificationRepository::save)
            .associateBy { it.id }

        // when
        val result = notificationRepository.findAllByTargetUserIdIn(usersIds)

        // then
        assertThat(result)
            .isNotEmpty
            .allSatisfy(Consumer { projection ->
                val currentId = projection.id
                val currentUserId = projection.targetUserId

                assertThat(notifications[currentId])
                    .describedAs("notification: $currentId, user id: $currentUserId")
                    .isNotNull
                    .satisfies(
                        Consumer {
                            assertThat(projection.id)
                                .describedAs("id")
                                .isEqualTo(it?.id)
                        },
                        {
                            assertThat(projection.targetUserId)
                                .describedAs("targetUserId")
                                .isEqualTo(it?.targetUserId)
                        },
                        {
                            assertThat(projection.message)
                                .describedAs("message")
                                .isEqualTo(it?.message)
                        },
                        {
                            assertThat(projection.severity)
                                .describedAs("severity")
                                .isEqualTo(it?.severity)
                        }
                    )
            })
    }

    private fun createUsers() = (0..5).map(::createPersistedTestUser)

    private fun createNotifications(targetUserId: Long) = (0..3).map {
        Notification(
            sourceUserId = -1,
            targetUserId = targetUserId,
            message = "test-message-$it",
            severity = "INFO",
            timestamp = LocalDateTime.now().truncatedTo(ChronoUnit.MILLIS),
        )
    }
}
